src/QsCompiler/CompilationManager/EditorCompletion.cs

```diff
@@ -0,0 +1,686 @@
+        private static FragmentRelationship GetFragmentAtOrBefore(
+            FileContentManager file,
+            Position position,
+            out CodeFragment.TokenIndex index,
+            out CodeFragment fragment)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+
+            var lineNumber = position.Line;
+            var tokens = file.GetTokenizedLine(lineNumber);
+
+            // If the current line is empty, find the last non-empty line before it.
+            while (tokens.IsEmpty && lineNumber > 0)
+                tokens = file.GetTokenizedLine(--lineNumber);
+
+            var indexNumber = tokens.TakeWhile(ContextBuilder.TokensUpTo(position)).Count() - 1;
+            if (indexNumber == -1)
+            {
+                index = null;
+                fragment = null;
+                return FragmentRelationship.Missing;
+            }
+
+            index = new CodeFragment.TokenIndex(file, lineNumber, indexNumber);
+            fragment = index.GetFragment();
+            if (fragment.FollowedBy != CodeFragment.MissingDelimiter &&
+                GetDelimitingCharPosition(file, fragment).IsSmallerThan(position))
+            {
+                switch (fragment.FollowedBy)
+                {
+                    case '}': return FragmentRelationship.InnerScope;
+                    case '{': return FragmentRelationship.EnclosingScope;
+                    case ';': return FragmentRelationship.SameScope;
+                }
+            }
+            return FragmentRelationship.Self;
+        }
+
+        /// <summary>
+        /// Returns the position of the delimiting character that follows the given code fragment.
+        /// </summary>
+        /// <exception cref="ArgumentException">Thrown when the code fragment has a missing delimiter.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when any argument is null.</exception>
+        private static Position GetDelimitingCharPosition(FileContentManager file, CodeFragment fragment)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (fragment == null)
+                throw new ArgumentNullException(nameof(fragment));
+            if (fragment.FollowedBy == CodeFragment.MissingDelimiter)
+                throw new ArgumentException("Code fragment has a missing delimiter", nameof(fragment));
+
+            var end = fragment.GetRange().End;
+            for (var lineNumber = end.Line; lineNumber < file.NrLines(); lineNumber++)
+            {
+                var start = lineNumber == end.Line ? end.Character : 0;
+                var line = file.GetLine(lineNumber);
+                var index = line.FindInCode(s => s.IndexOf(fragment.FollowedBy), start, line.Text.Length - start);
```

Comment:
So, I represent a missing delimiter by the char @, which may occur in the code for valid reasons. You need to ask the fragment is it's delimiter is missing first and special case on that.

Reply:
This method throws an exception if the delimiter is missing (line 668) - I think that should handle it?